### PR TITLE
Add support to dynamically resize threadpools size

### PR DESCRIPTION
**Pull request by gbbafna**
### Description

Adds capability to change thread pool sizes for all threadpools defined in core

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]

### Check List
- [x] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/internalClusterTest/java/org/opensearch/cluster/settings/ClusterSettingsIT.java [hunk 1]
```diff
@@ -380,6 +380,35 @@ public void testMissingUnits() {
...
+        transientSettings = Settings.builder().put(key1, "1").build();
+        persistentSettings = Settings.builder().put(key2, "5").build();
+        client().admin()
+            .cluster()
+            .prepareUpdateSettings()
+            .setTransientSettings(transientSettings)
+            .setPersistentSettings(persistentSettings)
+            .get();
+    }
+
```
Comment (Bukhtawar) [inline server/src/internalClusterTest/java/org/opensearch/cluster/settings/ClusterSettingsIT.java:411]: do we need to perform a get and re-assert the change?

File: server/src/main/java/org/opensearch/threadpool/ThreadPool.java [hunk 2]
```diff
@@ -222,10 +228,99 @@ public Collection<ExecutorBuilder> builders() {
...
     }
 
+    public void setThreadPool(Settings tpSettings) {
+        Map<String, Settings> tpGroups = tpSettings.getAsGroups();
+        for (Map.Entry<String, Settings> entry : tpGroups.entrySet()) {
+            String tpName = entry.getKey();
+            Settings tpGroup = entry.getValue();
+            ExecutorHolder holder = executors.get(tpName);
+            assert holder.executor instanceof OpenSearchThreadPoolExecutor;
+            OpenSearchThreadPoolExecutor o = (OpenSearchThreadPoolExecutor) holder.executor;
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/threadpool/ThreadPool.java:248]: will rename it .

File: server/src/main/java/org/opensearch/node/Node.java [hunk 3]
```diff
@@ -624,6 +624,7 @@ protected Node(
                 additionalSettingsFilter,
                 settingsUpgraders
             );
+            threadPool.setClusterSettings(settingsModule.getClusterSettings());
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/node/Node.java:627]: Is there any way we can avoid the cyclic dependency of these 2 classes one each other?

Comment (ashking94) [inline server/src/main/java/org/opensearch/node/Node.java:627]: Is there a possibility to put the cluster settings in the ctor of threadpool class?

File: server/src/main/java/org/opensearch/threadpool/ThreadPool.java [hunk 4]
```diff
@@ -222,10 +228,99 @@ public Collection<ExecutorBuilder> builders() {
         Setting.Property.NodeScope
     );
 
+    public static final Setting<Settings> CLUSTER_THREAD_POOL_SIZE_SETTING = Setting.groupSetting(
+        "cluster.thread_pool.",
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/threadpool/ThreadPool.java:232]: Why does this look incomplete?

File: server/src/main/java/org/opensearch/threadpool/ThreadPool.java [hunk 5]
```diff
@@ -222,10 +228,99 @@ public Collection<ExecutorBuilder> builders() {
...
+            }
+        }
+    }
+
+    public void setClusterSettings(ClusterSettings clusterSettings) {
+        this.clusterSettings = clusterSettings;
+        this.clusterSettings.addSettingsUpdateConsumer(CLUSTER_THREAD_POOL_SIZE_SETTING, this::setThreadPool, this::validateSetting);
+    }
+
+    private void validateSetting(Settings tpSettings) {
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/threadpool/ThreadPool.java:282]: Does this run first before each time we update a threadpool cluster setting?

File: server/src/main/java/org/opensearch/threadpool/ThreadPool.java [hunk 6]
```diff
@@ -222,10 +228,99 @@ public Collection<ExecutorBuilder> builders() {
...
+        Map<String, Settings> tpGroups = tpSettings.getAsGroups();
+        for (Map.Entry<String, Settings> entry : tpGroups.entrySet()) {
+            String tpName = entry.getKey();
+            if (THREAD_POOL_TYPES.get(tpName) == null) {
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/threadpool/ThreadPool.java:286]: nit - `if (THREAD_POOL_TYPES.containsKey(tpName))`?

File: server/src/main/java/org/opensearch/threadpool/ThreadPool.java [hunk 7]
```diff
@@ -222,10 +228,99 @@ public Collection<ExecutorBuilder> builders() {
...
+        }
+    }
+
+    public void setClusterSettings(ClusterSettings clusterSettings) {
+        this.clusterSettings = clusterSettings;
+        this.clusterSettings.addSettingsUpdateConsumer(CLUSTER_THREAD_POOL_SIZE_SETTING, this::setThreadPool, this::validateSetting);
+    }
+
+    private void validateSetting(Settings tpSettings) {
+        Map<String, Settings> tpGroups = tpSettings.getAsGroups();
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/threadpool/ThreadPool.java:283]: I believe this would give a map of threadpool name to thread pool settings? Can we add documentation as well on how would the threadpool settings need to be modified?

File: server/src/main/java/org/opensearch/threadpool/ThreadPool.java [hunk 8]
```diff
@@ -222,10 +228,99 @@ public Collection<ExecutorBuilder> builders() {
...
+            if (THREAD_POOL_TYPES.get(tpName) == null) {
+                throw new IllegalArgumentException("illegal thread_pool name : " + tpName);
+            }
+            Settings tpGroup = entry.getValue();
+            ExecutorHolder holder = executors.get(tpName);
+            assert holder.executor instanceof OpenSearchThreadPoolExecutor;
+            OpenSearchThreadPoolExecutor o = (OpenSearchThreadPoolExecutor) holder.executor;
+            if (holder.info.type == ThreadPoolType.SCALING) {
+                Set<String> so = new HashSet<>(Arrays.asList("max", "core"));
+                if (tpGroup.keySet().stream().allMatch(so::contains) == false) {
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/threadpool/ThreadPool.java:295]: this can be replaced with `so.containsAll(tpGroup.keySet())` for ease of reading. Also, do we need to throw exception? Or can we just log and move on?

File: server/src/main/java/org/opensearch/threadpool/ThreadPool.java [hunk 9]
```diff
@@ -222,10 +228,99 @@ public Collection<ExecutorBuilder> builders() {
...
+            String tpName = entry.getKey();
+            if (THREAD_POOL_TYPES.get(tpName) == null) {
+                throw new IllegalArgumentException("illegal thread_pool name : " + tpName);
+            }
+            Settings tpGroup = entry.getValue();
+            ExecutorHolder holder = executors.get(tpName);
+            assert holder.executor instanceof OpenSearchThreadPoolExecutor;
+            OpenSearchThreadPoolExecutor o = (OpenSearchThreadPoolExecutor) holder.executor;
+            if (holder.info.type == ThreadPoolType.SCALING) {
+                Set<String> so = new HashSet<>(Arrays.asList("max", "core"));
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/threadpool/ThreadPool.java:294]: Can we please add this in a constant somewhere?

File: server/src/main/java/org/opensearch/threadpool/ThreadPool.java [hunk 10]
```diff
@@ -222,10 +228,99 @@ public Collection<ExecutorBuilder> builders() {
...
+                if (tpGroup.keySet().stream().allMatch(so::contains) == false) {
+                    throw new IllegalArgumentException(
+                        "illegal thread_pool config : " + tpGroup.keySet() + " should only have max and core"
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/threadpool/ThreadPool.java:297]: We can use the constant from https://github.com/opensearch-project/OpenSearch/pull/16236/files#r1795171859 here?

File: server/src/main/java/org/opensearch/threadpool/ThreadPool.java [hunk 11]
```diff
@@ -222,10 +228,99 @@ public Collection<ExecutorBuilder> builders() {
...
+                } else {
+                    int size = tpGroup.getAsInt("size", o.getMaximumPoolSize());
+                    if (size < 1) {
+                        throw new IllegalArgumentException(
+                            "illegal value for [cluster.thread_pool." + tpName + "], has to be positive value"
+                        );
+                    }
+                }
+            }
+        }
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/threadpool/ThreadPool.java:321]: Why do we need to handle this differently?

File: server/src/main/java/org/opensearch/threadpool/ThreadPool.java [hunk 12]
```diff
@@ -222,10 +228,99 @@ public Collection<ExecutorBuilder> builders() {
...
+                 If we are increasing ,max pool size has to be increased first
+                 This ensures that core pool is always smaller than max pool size .
+                 */
+                if (core < o.getCorePoolSize()) {
+                    o.setCorePoolSize(core);
+                    o.setMaximumPoolSize(max);
+                } else {
+                    o.setMaximumPoolSize(max);
+                    o.setCorePoolSize(core);
+                }
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/threadpool/ThreadPool.java:263]: Not very clear why is this required.

File: server/src/main/java/org/opensearch/threadpool/ThreadPool.java [hunk 13]
```diff
@@ -222,10 +228,99 @@ public Collection<ExecutorBuilder> builders() {
...
+            } else {
+                int size = tpGroup.getAsInt("size", o.getMaximumPoolSize());
+                if (size < o.getCorePoolSize()) {
+                    o.setCorePoolSize(size);
+                    o.setMaximumPoolSize(size);
+                } else {
+                    o.setMaximumPoolSize(size);
+                    o.setCorePoolSize(size);
+                }
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/threadpool/ThreadPool.java:272]: Can we remove the duplication?

**Review by ashking94 — CHANGES_REQUESTED** [review]
Overall, I love the idea to tune the threadpool size. This will give end users the option to get the best out of their opensearch cluster.

File: server/src/main/java/org/opensearch/threadpool/ThreadPool.java (hunk 4 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/threadpool/ThreadPool.java:232]: That is  how group settings are defined .

**Comment by jed326** [comment]
Thanks @gbbafna! I like the idea of dynamically tuning the threadpool size but I am worried there are some sharp edges for users with a cluster setting.

Today the threadpool size can be set per-node via the `opensearch.yml`, so in cases where there is a mix of hardware type across the cluster each node can have it's own configuration. This is pretty typical for the cluster manager nodes and the data nodes to be of different instance types, so the concern is that we could accidentally adversely affect one instance type by setting the cluster wide setting.

For example, if we increase the snapshots threadpool size across the cluster, that might improve the performance of taking a snapshot on the data nodes, but have adverse effects on cloning a snapshot on the cluster manager nodes.

**Comment by dbwiddis** [comment]
> Overall, I love the idea to tune the threadpool size. This will give end users the option to get the best out of their opensearch cluster.

I love the idea of tuning, but that also just adds a lot more complexity that barely any of us understand!  What *should* it be?  

Giving one threat pool more resources should ideally be balanced by a reduction in another, but how do we know?  Is there any way to have an overall "max threads across all these different pools" constraint to let people tweak up one pool as long as they constrain others?   Are there stats that let us measure how many times a pool is maximized?

File: server/src/main/java/org/opensearch/node/Node.java (hunk 3 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/node/Node.java:627]: clusterSettings depend on thread pool . So we can't put cluster  settings in the ctor of threadpool class.

File: server/src/main/java/org/opensearch/threadpool/ThreadPool.java (hunk 5 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/threadpool/ThreadPool.java:282]: Yes.

File: server/src/main/java/org/opensearch/threadpool/ThreadPool.java (hunk 7 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/threadpool/ThreadPool.java:283]: Ack .

File: server/src/main/java/org/opensearch/threadpool/ThreadPool.java (hunk 8 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/threadpool/ThreadPool.java:295]: Done. We need to throw exception as we can't apply that setting .

File: server/src/main/java/org/opensearch/threadpool/ThreadPool.java (hunk 9 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/threadpool/ThreadPool.java:294]: Done.

File: server/src/main/java/org/opensearch/threadpool/ThreadPool.java (hunk 11 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/threadpool/ThreadPool.java:321]: This is simplified now .

File: server/src/main/java/org/opensearch/threadpool/ThreadPool.java (hunk 13 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/threadpool/ThreadPool.java:272]: We need to make sure that `core > max` pool size at all the times. Hence we need to it this way only .

**Comment by gbbafna** [comment]
> Thanks @gbbafna! I like the idea of dynamically tuning the threadpool size but I am worried there are some sharp edges for users with a cluster setting.
> 
> Today the threadpool size can be set per-node via the `opensearch.yml`, so in cases where there is a mix of hardware type across the cluster each node can have it's own configuration. This is pretty typical for the cluster manager nodes and the data nodes to be of different instance types, so the concern is that we could accidentally adversely affect one instance type by setting the cluster wide setting.
> 
> For example, if we increase the snapshots threadpool size across the cluster, that might improve the performance of taking a snapshot on the data nodes, but have adverse effects on cloning a snapshot on the cluster manager nodes.

This are quite valid points @jed326, thanks for them . Definitely, there are sharp edges with this feature.  We would recommend only expert users to tune this up . Also since its dynamic, users can easily revert it as well on observation of degradation.  On a higher level , i would say we should have a different thread pool for cloning and create snapshot . Recently we introduced another threadpool for deletion as well to independently scale up and down. 

Also i would put in the documentation to not use this if you have a heterogeneous cluster . Practically , I have seen most of the clusters as homogeneous only .

**Comment by gbbafna** [comment]
> > Overall, I love the idea to tune the threadpool size. This will give end users the option to get the best out of their opensearch cluster.
> 
> I love the idea of tuning, but that also just adds a lot more complexity that barely any of us understand! What _should_ it be?
> 
> Giving one threat pool more resources should ideally be balanced by a reduction in another, but how do we know? 

Thanks @dbwiddis . I agree it is tricky to tune it and would document that only experts should tune it on a need basis and monitor for any degradation. 

`how do we know? ` - There would be some experimentation required to figure out what is best for a given configuration. 

>Is there any way to have an overall "max threads across all these different pools" constraint to let people tweak up one pool as long as they constrain others?

We can think about this if this will benefit our users .

>  Are there stats that let us measure how many times a pool is maximized?

Yes, we have [thread pool stats](https://opensearch.org/docs/latest/api-reference/nodes-apis/nodes-stats/#thread_pool) to figure that out .

File: server/src/internalClusterTest/java/org/opensearch/cluster/settings/ClusterSettingsIT.java [hunk 14]
```diff
@@ -380,6 +380,100 @@ public void testMissingUnits() {
...
+                .cluster()
+                .prepareUpdateSettings()
+                .setTransientSettings(
+                    Settings.builder().put("cluster.thread_pool.snapshot.core", "2").put("cluster.thread_pool.snapshot.max", "1").build()
+                )
+                .get();
+            fail("bogus value");
+        } catch (IllegalArgumentException ex) {
+            assertEquals(ex.getCause().getMessage(), "core threadpool size cannot be greater than max");
+        }
```
Comment (ashking94) [inline server/src/internalClusterTest/java/org/opensearch/cluster/settings/ClusterSettingsIT.java:432]: Should we also have tests for combination of default value & transient settings value of core/max settings?

File: server/src/internalClusterTest/java/org/opensearch/cluster/settings/ClusterSettingsIT.java [hunk 15]
```diff
@@ -380,6 +380,100 @@ public void testMissingUnits() {
...
+
+        // Scaling threadpool - happy case
+        ClusterUpdateSettingsResponse clusterUpdateSettingsResponse = client().admin()
+            .cluster()
+            .prepareUpdateSettings()
+            .setTransientSettings(Settings.builder().put("cluster.thread_pool.snapshot.max", "1").build())
+            .setPersistentSettings(Settings.builder().put("cluster.thread_pool.snapshot.max", "5").build())
```
Comment (ashking94) [inline server/src/internalClusterTest/java/org/opensearch/cluster/settings/ClusterSettingsIT.java:439]: Unclear what are we testing here?

File: server/src/main/java/org/opensearch/threadpool/ThreadPool.java [hunk 16]
```diff
@@ -403,6 +416,94 @@ public Info info(String name) {
         return holder.info;
     }
 
+    public void registerClusterSettingsListeners(ClusterSettings clusterSettings) {
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/threadpool/ThreadPool.java:419]: Looks better now

**Review by ashking94 — COMMENTED** [review]
LGTM overall, have left some comments.

File: server/src/main/java/org/opensearch/threadpool/ThreadPool.java [hunk 17]
```diff
@@ -211,6 +215,8 @@ public static ThreadPoolType fromType(String type) {
 
     private final ScheduledThreadPoolExecutor scheduler;
 
+    private ClusterSettings clusterSettings = null;
```
[the remark this answers is not in the document]
Comment (ashking94) [reply server/src/main/java/org/opensearch/threadpool/ThreadPool.java]: I don't think that we require this anymore.

File: server/src/internalClusterTest/java/org/opensearch/cluster/settings/ClusterSettingsIT.java (hunk 15 above)
Comment (gbbafna) [reply server/src/internalClusterTest/java/org/opensearch/cluster/settings/ClusterSettingsIT.java:439]: Wanted to verify transient setting override persistent settings and we correctly validate `max >=core` . Modified the settings to do same now .

**Review by ashking94 — APPROVED** [review]
